src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -55,71 +60,221 @@ protected override string CreateRequestId()
         protected override MessageBody CreateMessageBody()
             => Http2MessageBody.For(HttpRequestHeaders, this);
 
+        // Compare to Http1Connection.OnStartLine
         protected override bool TryParseRequest(ReadResult result, out bool endConnection)
         {
             // We don't need any of the parameters because we don't implement BeginRead to actually
             // do the reading from a pipeline, nor do we use endConnection to report connection-level errors.
+            endConnection = !TryValidatePseudoHeaders();
+            return true;
+        }
+
+        private bool TryValidatePseudoHeaders()
```

Comment:
Is there now way to do this before we turn them into strings?

Reply:
Not without significant changes to the decoder and header collections.
1. Make these all Known headers https://github.com/aspnet/KestrelHttpServer/issues/2261
2. Make the decoder return raw bytes for Known headers
3. Do validation on the raw bytes, then decode
4. Lazy decode any remaining known headers

The :path is the only field this will make a perf difference for on valid requests, it's the only one transformed.